src/app/outbound.rs

```diff
@@ -157,85 +157,111 @@ impl tap::Inspect for Endpoint {
-            Resolve(resolve)
+        pub fn new(resolver: R) -> Self {
+            Resolver(resolver)
         }
     }
 
-    impl<R> resolve::Resolve<DstAddr> for Resolve<R>
+    impl<R> resolve::Resolve<DstAddr> for Resolver<R>
     where
         R: resolve::Resolve<NameAddr, Endpoint = Metadata>,
     {
         type Endpoint = Endpoint;
         type Resolution = Resolution<R::Resolution>;
-        type Future = Resolution<R::Future>;
+        type Future = ResolveFuture<R::Future>;
 
         fn resolve(&self, dst: &DstAddr) -> Self::Future {
-            let resolving = match dst.as_ref() {
-                Addr::Name(ref name) => Resolving::Name(name.clone(), self.0.resolve(&name)),
-                Addr::Socket(ref addr) => Resolving::Addr(Some(*addr)),
-            };
-
-            Resolution {
-                http_settings: dst.http_settings,
-                resolving,
+            match dst.as_ref() {
+                Addr::Name(ref name) => {
+                    ResolveFuture::new(dst.http_settings, name.clone(), self.0.resolve(&name))
+                }
+                Addr::Socket(_) => ResolveFuture::unresolvable(),
             }
         }
     }
 
-    // === impl Resolution ===
+    // ===== impl ResolveFuture =====
+
+    impl<F> ResolveFuture<F> {
+        fn new(http_settings: settings::Settings, name: NameAddr, resolution: F) -> Self {
+            Self {
+                resolving: Some(Resolution {
+                    http_settings,
+                    name,
+                    resolution,
+                }),
+            }
+        }
+
+        fn unresolvable() -> Self {
+            Self { resolving: None }
+        }
+    }
 
-    impl<F: Future> Future for Resolution<F> {
+    impl<F: Future> Future for ResolveFuture<F>
+    where
+        F: Future,
+    {
         type Item = Resolution<F::Item>;
-        type Error = F::Error;
+        type Error = Unresolvable;
+
         fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-            let resolving = match self.resolving {
-                Resolving::Name(ref name, ref mut f) => {
-                    let res = try_ready!(f.poll());
-                    // TODO: get rid of unnecessary arc bumps?
-                    Resolving::Name(name.clone(), res)
-                }
-                Resolving::Addr(a) => Resolving::Addr(a),
-            };
-            Ok(Async::Ready(Resolution {
-                resolving,
-                // TODO: get rid of unnecessary clone
-                http_settings: self.http_settings.clone(),
-            }))
+            if let Some(Resolution {
+                http_settings,
+                ref name,
+                ref mut resolution,
+            }) = self.resolving
+            {
+                let resolution = match resolution.poll() {
+                    Ok(Async::Ready(resolution)) => resolution,
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Err(_) => {
```

Comment:
Do we know that `Unresolvable` is the only error that the `control::destination` resolver will return?